File: examples/typescript/04_networks.ts
/**
 * 04_networks.ts — Create a network, join it from a second agent, broadcast.
 *
 * Run:
 *   AGENT_B_ADDR=unix://$HOME/.moltmesh/agent_b.sock npx tsx 04_networks.ts
 */
import { A2AClient } from "../../sdk/typescript/openclaw-plugin/src/client.js";

const AGENT_A_ADDR = process.env["AGENT_A_ADDR"] ?? "";
const AGENT_B_ADDR = process.env["AGENT_B_ADDR"] ?? `unix://${process.env["HOME"]}/.moltmesh/agent_b.sock`;

async function listenOne(client: A2AClient, networkId: string): Promise<void> {
  console.log(`[agent_b] subscribing to network ${networkId.slice(0, 8)}...`);
  for await (const msg of client.subscribeNetwork(networkId)) {
    console.log(`[agent_b] broadcast: ${Buffer.from(msg.payload).toString("utf8")}`);
    break;
  }
}

async function main() {
  const a = new A2AClient(AGENT_A_ADDR);
  const b = new A2AClient(AGENT_B_ADDR);

  const net = await a.createNetwork("demo-group");
  console.log(`Network created: ${net.id}  name=${net.name}`);

  await b.joinNetwork(net.id);
  console.log(`Agent B joined network ${net.id.slice(0, 8)}`);

  const listenPromise = listenOne(b, net.id);

  await a.broadcastNetwork(net.id, "Hello network!");
  console.log(`[agent_a] broadcast sent`);

  await listenPromise;

  const netsB = await b.listNetworks();
  console.log(`\nAgent B networks: ${netsB.map(n => n.name).join(", ")}`);

  a.close();
  b.close();
}

main().catch(console.error);

File: examples/python/04_networks.py
"""
04_networks.py — Create a group, have a second agent join, and broadcast.

Run:
    AGENT_B_ADDR=unix://$HOME/.moltmesh/agent_b.sock python 04_networks.py
"""
import os
import threading
from moltmesh import A2AClient

AGENT_A_ADDR = os.getenv("AGENT_A_ADDR", "")
AGENT_B_ADDR = os.getenv("AGENT_B_ADDR", f"unix://{os.path.expanduser('~')}/.moltmesh/agent_b.sock")

def listen(client: A2AClient, network_id: str):
    print(f"[agent_b] subscribing to network {network_id[:8]}...")
    for msg in client.subscribe_network(network_id):
        print(f"[agent_b] broadcast: {msg.payload.decode()}")
        break  # receive one message then exit

def main():
    a = A2AClient(AGENT_A_ADDR).connect()
    b = A2AClient(AGENT_B_ADDR).connect()

    net = a.create_network("demo-group")
    print(f"Network created: {net.id}  name={net.name}")

    b.join_network(net.id)
    print(f"Agent B joined network {net.id[:8]}")

    t = threading.Thread(target=listen, args=(b, net.id), daemon=True)
    t.start()

    a.broadcast_network(net.id, "Hello network!")
    print(f"[agent_a] broadcast sent")

    t.join(timeout=5)

    nets_b = b.list_networks()
    print(f"\nAgent B networks: {[n.name for n in nets_b]}")

    a.close()
    b.close()

if __name__ == "__main__":
    main()
